widget/tree.go

```diff
@@ -52,6 +52,7 @@ type Tree struct {
 	offset        fyne.Position
 	open          map[TreeNodeID]bool
 	scroller      *widget.Scroll
+	content       *treeContent
```

Comment:
This does kind-of feel like a band-aid fix since conceptually, treeRenderer should own the treeContent widget. I'd rather get to the bottom of how CreateRenderer is actually called twice in a race condition. Maybe @andydotxyz has ideas?

Reply:
I'm certainly in favor of a more generic fix if there is one.  Here are the two backtraces that both called CreateRenderer:

```go
fyne.io/fyne/v2/widget.newTreeContent(tree.go:580)
fyne.io/fyne/v2/widget.(*Tree).CreateRenderer(tree.go:142)
fyne.io/fyne/v2/internal/cache.Renderer(widget.go:33)
fyne.io/fyne/v2/widget.(*BaseWidget).MinSize(widget.go:74)
fyne.io/fyne/v2/widget.(*Tree).MinSize(tree.go:196)
fyne.io/fyne/v2/layout.(*borderLayout).MinSize(borderlayout.go:79)
fyne.io/fyne/v2.(*Container).MinSize(container.go:90)
fyne.io/fyne/v2/internal/driver/glfw.(*glCanvas).SetContent(canvas.go:158)
fyne.io/fyne/v2/internal/driver/glfw.(*window).SetContent(window.go:256)
main.main(main.go:83)
runtime.main(proc.go:271)
runtime.goexit(asm_amd64.s:1695)
runtime.newproc(:0)


fyne.io/fyne/v2/widget.newTreeContent(tree.go:580)
fyne.io/fyne/v2/widget.(*Tree).CreateRenderer(tree.go:142)
fyne.io/fyne/v2/internal/cache.Renderer(widget.go:33)
fyne.io/fyne/v2/widget.(*BaseWidget).Refresh(widget.go:136)
fyne.io/fyne/v2/widget.(*BaseWidget).Refresh-fm(:0)
fyne.io/fyne/v2/data/binding.(*listener).DataChanged(binding.go:56)
fyne.io/fyne/v2/data/binding.DataListener.DataChanged-fm(:0)
fyne.io/fyne/v2/data/binding.queueItem.func1.1(queue.go:19)
runtime.goexit(asm_amd64.s:1695)
fyne.io/fyne/v2/data/binding.queueItem.func1(queue.go:17)
```

The code for getting the renderer clearly doesn't hold a lock when doing `CreateRenderer` which feels like the underlying issue here:

```go
        renderersLock.RLock()
	rinfo, ok := renderers[wid]
	renderersLock.RUnlock()
	if !ok {
		rinfo = &rendererInfo{renderer: wid.CreateRenderer()}
		renderersLock.Lock()
		renderers[wid] = rinfo
		renderersLock.Unlock()
	}
```

But I experimented a bit with locking there and found that Renderers can create other renderers which would create deadlock in some places.  I believe we either need to find a way to lock when calling CreateRenderer (at least for a single widget instance), or ensure all CreateRenderer methods have no side affects.  It seems more difficult to enforce that CreateRenderer doesn't have side affects BUT users are supposed to expect them to be destroyed and recreated - so I question if side affects should be avoided anyways.

Reply:
Oh, I think the problem may have to do with bindings as they run on a separate thread IIRC. Yeah we might need @andydotxyz to suggest the best plan for a fix

Reply:
I see this is closed but I wanted to comment as pinged incase it helps.
Locking access to renderers whilst one is created wouldn't be the right way - but locking on a specific key would be OK. However as this seems to step from a binding issue then it will likely be resolved by our threading changes for 2.6.

The overall question of when is a render created still applies, however we have no guarantee that `CreateRenderer` is called only once (in fact we say it may be called many times). As long as the new renderer replaces the old which will be garbage collected then I think everything should continue to work as planned?